lib/srv/db/common/databaseobjectimportrule/apply.go

```diff
@@ -0,0 +1,281 @@
+			case "database":
+				return spec.GetDatabase(), nil
+			case "schema":
+				return spec.GetSchema(), nil
+			case "name":
+				return spec.GetName(), nil
+			}
+
+			return "", trace.NotFound("key %v not found", key)
+		}),
+	}
+
+	parser, err := typical.NewParser[evaluationEnv, string](typical.ParserSpec{Variables: envVar})
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	expr, err := parser.Parse(e.text)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	text, err := expr.Evaluate(evaluationEnv{})
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	return text, nil
+}
+
+var reVariable = regexp.MustCompile(
+	// prefix is anything that is not { or }
+	`^(?P<prefix>[^}{]*)` +
+		// variable is anything in brackets {{}} that is not { or }
+		`{{(?P<expression>\s*[^}{]*\s*)}}` +
+		// suffix is anything that is not { or }
+		`(?P<suffix>[^}{]*)$`,
+)
+
+// splitExpression splits the template into several parts, to be evaluated separately.
+func splitExpression(value string) ([]eval, error) {
+	match := reVariable.FindStringSubmatch(value)
+	if len(match) == 0 {
+		if strings.Contains(value, "{{") || strings.Contains(value, "}}") {
+			return nil, trace.BadParameter(
+				"%q is using template brackets '{{' or '}}', however expression does not parse, make sure the format is {{expression}}",
+				value,
+			)
+		}
+		return []eval{literal{text: strings.TrimSpace(value)}}, nil
+	}
+
+	return []eval{
+		literal{text: strings.TrimLeftFunc(match[1], unicode.IsSpace)},
+		expression{text: match[2]},
+		literal{text: strings.TrimRightFunc(match[3], unicode.IsSpace)},
+	}, nil
+}
+
+func evalTemplate(template string, spec *dbobjectv1.DatabaseObjectSpec) (string, error) {
+	chunks, err := splitExpression(template)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	var sb strings.Builder
+
+	for _, chunk := range chunks {
+		text, err := chunk.eval(spec)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		sb.WriteString(text)
+	}
+
+	return sb.String(), nil
+}
+
+func applyMappingToObject(mapping *dbobjectimportrulev1.DatabaseObjectImportRuleMapping, spec *dbobjectv1.DatabaseObjectSpec, labels map[string]string) (bool, error) {
+	// the matching is applied to the object spec; existing object labels does not matter
+	if !databaseObjectScopeMatch(mapping.GetScope(), spec) {
+		return false, nil
+	}
+	if !databaseObjectImportMatch(mapping.GetMatch(), spec) {
+		return false, nil
+	}
+
+	for key, value := range mapping.AddLabels {
+		out, err := evalTemplate(value, spec)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+		labels[key] = out
+	}
+
+	return true, nil
+}
+
+func matchPattern(pattern, value string) bool {
+	re, err := libutils.CompileExpression(pattern)
+	if err != nil {
+		return false
+	}
+	return re.MatchString(value)
+}
+
+func matchAny(patterns []string, value string) bool {
+	return utils.Any(patterns, func(pattern string) bool {
+		return matchPattern(pattern, value)
+	})
+}
```

Comment:
Well, that would be my preference too, but switching to `{{.name}}` would be inconsistent with other places in Teleport.

Reply:
I don't think that we are other places where a  mustache template is used in teleport so {{.name} vs {{name}} is just a UI decision. 


@xinding33 @r0mant  
Any thought/preference on following UI ? 

```yaml
mappings:
    - add_labels:
        database: '{{obj.database}}'
        kind: '{{obj.object_kind}}'
        name: '{{obj.name}}'
        protocol: '{{obj.protocol}}'
        schema: '{{obj.schema}}'
        service_name: '{{obj.database_service_name}}'
```    
 **VS**
```yaml
mappings:
    - add_labels:
        database: '{{.database}}'
        kind: '{{.object_kind}}'
        name: '{{.name}}'
        protocol: '{{.protocol}}'
        schema: '{{.schema}}'
        service_name: '{{.database_service_name}}'
```

Reply:
Hmm, there are a few? Most notably roles:

```
    db_users: ['{{email.local(external.email)}}']
    db_names: ['{{external.db_names}}']
    db_roles: ['{{external.db_roles}}']
    db_labels:
      'env': '{{regexp.replace(external.env, "^(staging)$", "$1")}}'
```

Reply:
> Hmm, there are a few? Most notably roles:

My bad. For me it quite weird that if we already have mustache language support in other places for DB flow we are creating the same thing from scratch instead of reusing already existed implementation.

Reply:
Ya, we had a few discussions on using existing package. @Tener actually started with it.
https://github.com/gravitational/teleport/pull/38630#discussion_r1507874496
https://github.com/gravitational/teleport/pull/38630#discussion_r1512977376

Reply:
Yeah, we are reusing as much from existing implementation as possible given the slightly different context and requirements.

Reply:
What would you say about  making simplest generic  implementation parse logic  and moving it into common package utils/parse  where the logic can be refused and it quite readable  ?

```go
func evalTemplate(value string, obj any) (string, error) {
	match := reVariable.FindStringSubmatch(value)
	if len(match) == 0 {
		if strings.Contains(value, "{{") || strings.Contains(value, "}}") {
			return "", trace.BadParameter(
				"%q is using template brackets '{{' or '}}', however expression does not parse, make sure the format is {{expression}}",
				value,
			)
		}
		return value, nil
	}
	prefix, value, suffix := strings.TrimLeftFunc(match[1], unicode.IsSpace), match[2], strings.TrimRightFunc(match[3], unicode.IsSpace)
	p, err := predicate.NewParser(predicate.Def{
		GetIdentifier: func(fields []string) (interface{}, error) {
			return predicate.GetFieldByTag(obj, teleport.JSON, fields)
		},
	})

	parsed, err := p.Parse(value)
	if err != nil {
		return "", trace.Wrap(err)
	}
	result, ok := parsed.(string)
	if !ok {
		return "", trace.BadParameter("expected string, got %T", parsed)
	}

	var sb strings.Builder
	sb.WriteString(prefix)
	sb.WriteString(result)
	sb.WriteString(suffix)
	return sb.String(), nil
}
```

Reply:
I think having a specific object to reference (here `obj`) is important for extensibility POV. That said, a minor adjustment in `EvalTemplate` restores that:

```golang
func EvalTemplate(value string, namespace map[string]any) (string, error) {
	match := reVariable.FindStringSubmatch(value)
	if len(match) == 0 {
		if strings.Contains(value, "{{") || strings.Contains(value, "}}") {
			return "", trace.BadParameter(
				"%q is using template brackets '{{' or '}}', however expression does not parse, make sure the format is {{expression}}",
				value,
			)
		}
		return value, nil
	}
	prefix, value, suffix := strings.TrimLeftFunc(match[1], unicode.IsSpace), match[2], strings.TrimRightFunc(match[3], unicode.IsSpace)
	p, err := predicate.NewParser(predicate.Def{
		GetIdentifier: func(fields []string) (interface{}, error) {
			obj, found := namespace[fields[0]]
			if !found {
				return nil, trace.NotFound("identifier %q is not defined", strings.Join(fields, "."))
			}
			return predicate.GetFieldByTag(obj, teleport.JSON, fields[1:])
		},
	})
       // ...
}
```

This is neat. 

I'm not convinced, however, that `EvalTemplate` should be added to `parse`. I've expressed my concerns previously (https://github.com/gravitational/teleport/pull/38630#discussion_r1513163706):

> The underlying library here is `typical`, which is also our own code. `parse` sounds general, but is truly about traits, and I think we should avoid adding extra features there. There is also [`lib/expression`](https://github.com/gravitational/teleport.e/tree/master/lib/expression) library in `teleport.e`, but it is also based on `typical`.
> 
> The only thing used from `parse` is `parse.SplitExpression`, which is a newly exposed utility function. I'm strongly considering breaking this dependency completely - revert the changes to `parse` and just reimplement this function locally; it isn't complex and this would remove this unnecessary dependency.

All in all, I just don't think adding a new utility function to `parse`, unused by anything related to traits, is the right call. Thoughts?

Reply:
> Personally I would move it to parse because we have already the reVariable defined in this packet and if first directory that I would check to see if I can reuse something before writing a custom parser :-)

> Nevertheless minimal version of EvalTemplate function implementation is small and if we descide to keep the obj reference we should keep it in db package.

I'm not super happy about copying the regexp, but creating this dependency on `parse` doesn't feel good either. Right now, I believe that avoiding the dependency is better. This choice is also supported by various other uses of `predicate.NewParser` and `typical.NewParser`: the overall design choice seems to be to do the parsing independently on relevant places.

> I thought that this was a side effect decision to support `{{obj.object_kind}}` instead of `{{object_kind}}`

I haven't considered using `{{object_kind}}` at all, as it goes against the extensibility goal, but I'm not sure if `{{object_kind}}` would work any better; `internal` and `external` are hard-coded in relevant places.